src/Avalonia.Base/Utilities/AvaloniaPropertyValueStore.cs

```diff
@@ -8,166 +7,197 @@ namespace Avalonia.Utilities
             }
 
-            value = _entries[index].Value;
+            value = _entries![entryIndex.Index].Value;
+            
             return true;
         }
 
-        public void AddValue(AvaloniaProperty property, TValue value)
+        private void InsertEntry(Entry entry, int entryIndex)
         {
-            Entry[] entries = new Entry[_entries.Length + 1];
-
-            for (int i = 0; i < _entries.Length; ++i)
+            if (_entryCount > 0)
             {
-                if (_entries[i].PropertyId > property.Id)
+                if (_entryCount == _entries!.Length)
                 {
-                    if (i > 0)
+                    // We want to have more aggressive resizing when initializing.
+                    var growthFactor = IsDuringInitialization ? 2.0 : 1.2;
+                    
+                    var newSize = (int)(_entryCount * growthFactor);
+
+                    if (newSize == _entryCount)
                     {
-                        Array.Copy(_entries, 0, entries, 0, i);
+                        newSize++;
                     }
 
-                    entries[i] = new Entry { PropertyId = property.Id, Value = value };
-                    Array.Copy(_entries, i, entries, i + 1, _entries.Length - i);
-                    break;
+                    var destEntries = new Entry[newSize];
+
+                    Array.Copy(_entries, 0, destEntries, 0, entryIndex);
+
+                    destEntries[entryIndex] = entry;
+                    
+                    Array.Copy(_entries, entryIndex, destEntries, entryIndex + 1, _entryCount - entryIndex);
+                    
+                    _entries = destEntries;
                 }
+                else
+                {
+                    Array.Copy(
+                        _entries,
+                        entryIndex,
+                        _entries,
+                        entryIndex + 1,
+                        _entryCount - entryIndex);
+
+                    _entries[entryIndex] = entry;
+                }
+            }
+            else
+            {
+                if (_entries is null)
+                {
+                    _entries = new Entry[InitialSize];
+                }
+                
+                _entries[0] = entry;
             }
 
-            _entries = entries;
+            _entryCount++;
+        }
+
+        public void AddValue(AvaloniaProperty property, TValue value)
+        {
+            var propertyId = property.Id;
+            var index = LookupEntry(propertyId);
+
+            InsertEntry(new Entry(propertyId, value), index.Index);
         }
 
         public void SetValue(AvaloniaProperty property, TValue value)
         {
-            _entries[TryFindEntry(property.Id).Item1].Value = value;
+            var propertyId = property.Id;
+            var entryIndex = LookupEntry(propertyId);
+            
+            _entries![entryIndex.Index] = new Entry(propertyId, value);
```

Comment:
According to previous implementation and WPF - it should crash and burn when lookup entry fails since that means API was misused. Could add an assert in debug since this shouldn't happen at all unless there is a framework bug.

Reply:
Dictionary `[]` operator takes a key object and not index.